rotator/rotator.go

```diff
@@ -102,34 +110,54 @@ func (r *Rotator) rotate() error {
 		}
 	}
 
-	arcPath := fmt.Sprintf("%s.%d.gz", r.filename, maxNum+1)
-
-	arc, err := os.OpenFile(arcPath, os.O_CREATE|os.O_EXCL|os.O_WRONLY, 0644)
+	err = r.out.Close()
 	if err != nil {
 		return err
 	}
-
-	r.out.Seek(0, os.SEEK_SET)
-	z := gzip.NewWriter(arc)
-	if _, err = io.Copy(z, r.out); err != nil {
-		return err
-	}
-	if err = z.Close(); err != nil {
+	rotname := fmt.Sprintf("%s.%d", r.filename, maxNum+1)
+	err = os.Rename(r.filename, rotname)
+	if err != nil {
 		return err
 	}
-	if err := arc.Close(); err != nil {
+	r.out, err = os.OpenFile(r.filename, os.O_CREATE|os.O_RDWR, 0644)
+	if err != nil {
 		return err
 	}
-	if err = r.out.Close(); err != nil {
+	r.size = 0
+
+	r.wg.Add(1)
+	go func() {
+		compress(rotname)
+		r.wg.Done()
+	}()
+
+	return nil
+}
+
+func compress(name string) (err error) {
+	f, err := os.Open(name)
+	if err != nil {
 		return err
 	}
 
-	r.out, err = os.OpenFile(r.filename, os.O_CREATE|os.O_TRUNC|os.O_RDWR, 0644)
+	defer func() {
+		f.Close()
+		if err == nil {
```

Comment:
no, it should be the closed-over return error of the compress function, although i can see that it is not obvious without a different variable name.  Idea is to only remove the original file if the compression step was successful.